packages/ember-glimmer/lib/syntax/curly-component.js

```diff
@@ -105,6 +97,38 @@ class CurlyComponentManager {
     return bucket;
   }
 
+  ensureCompilable(definition, bucket, env) {
+    if (definition.template) {
+      return definition;
+    }
+
+    let { component } = bucket;
+    let TemplateFactory = component.layout;
+    let template;
+    let owner = getOwner(component);
+
+    // If the user has supplied their own layout property we must
+    // create an instance of the wrapper class with the env to go back into
+    // the runtime compiler phase :(
+    if (TemplateFactory) {
+      template = new TemplateFactory(env);
```

Comment:
We should move the id to the template() function we compile the modules with and give the id to the factory, so we can cache the factory instance on the late bound component definition.  We also should assume that the template wasn't looked up via the container.